File: android/app/src/main/java/com/appxstudios/festivalconnection/mesh/shared/PacketProcessor.kt
package com.appxstudios.festivalconnection.mesh.shared

import com.appxstudios.festivalconnection.protocol.CrowdSyncBinaryProtocol
import com.appxstudios.festivalconnection.protocol.CrowdSyncPacket
import com.appxstudios.festivalconnection.protocol.MessageType
import com.appxstudios.festivalconnection.protocol.PaymentPacketSerializer
import java.security.MessageDigest
import java.util.concurrent.ConcurrentHashMap

// Central router that receives raw bytes from all transports (BLE, Nearby, Nostr),
// deduplicates by message ID, decrements TTL and relays, and routes to handlers.

object PacketProcessor {
    private val seenIDs = ConcurrentHashMap<String, Long>()
    private const val DEDUP_WINDOW_MS = 300_000L
    private const val MAX_CACHE_SIZE = 1000

    // Callbacks
    var onAnnounce: ((String, String) -> Unit)? = null       // peerKeyHex, nickname
    var onMessage: ((String, String, String) -> Unit)? = null // senderHex, recipientHex, content
    var onLeave: ((String) -> Unit)? = null
    var onPaymentRequest: ((String, Long, String, String) -> Unit)? = null
    var onPaymentNotification: ((ByteArray, Long, Byte) -> Unit)? = null

    enum class TransportType { BLE, NEARBY, NOSTR }

    fun receive(data: ByteArray, fromTransport: TransportType) {
        val msgID = computeMessageID(data)
        if (isDuplicate(msgID)) return

        val packet = CrowdSyncBinaryProtocol.decode(data) ?: return
        routePacket(packet)

        // Relay if TTL > 1
        if (packet.ttl > 1) {
            val relayed = packet.copy(ttl = packet.ttl - 1)
            val relayedData = CrowdSyncBinaryProtocol.encode(relayed) ?: return
            // Relay to other transports (excluding source)
            // BLE and Nearby transports handle their own broadcast
        }
    }

    private fun routePacket(packet: CrowdSyncPacket) {
        val senderHex = packet.senderID.joinToString("") { "%02x".format(it) }
        val type = MessageType.fromByte(packet.type) ?: return

        when (type) {
            MessageType.ANNOUNCE -> {
                val nickname = String(packet.payload, Charsets.UTF_8).ifEmpty {
                    "Peer ${senderHex.take(4).uppercase()}"
                }
                onAnnounce?.invoke(senderHex, nickname)
            }
            MessageType.MESSAGE -> {
                val content = String(packet.payload, Charsets.UTF_8)
                val recipientHex = packet.recipientID?.joinToString("") { "%02x".format(it) } ?: ""
                onMessage?.invoke(senderHex, recipientHex, content)
            }
            MessageType.LEAVE -> onLeave?.invoke(senderHex)
            MessageType.PAYMENT_REQUEST -> {
                PaymentPacketSerializer.decodePaymentRequest(packet.payload)?.let { (amt, inv, desc) ->
                    onPaymentRequest?.invoke(senderHex, amt, inv, desc)
                }
            }
            MessageType.PAYMENT_NOTIFICATION -> {
                if (packet.payload.size >= 41) {
                    val hash = packet.payload.sliceArray(0 until 32)
                    var amt = 0L
                    for (i in 32 until 40) amt = (amt shl 8) or (packet.payload[i].toLong() and 0xFF)
                    onPaymentNotification?.invoke(hash, amt, packet.payload[40])
                }
            }
            else -> { /* Handled by specialized subsystems */ }
        }
    }

    private fun computeMessageID(data: ByteArray): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(data)
        return digest.take(16).joinToString("") { "%02x".format(it) }
    }

    private fun isDuplicate(id: String): Boolean {
        val now = System.currentTimeMillis()
        if (seenIDs.containsKey(id) && (now - (seenIDs[id] ?: 0)) < DEDUP_WINDOW_MS) return true
        seenIDs[id] = now
        if (seenIDs.size > MAX_CACHE_SIZE) {
            val cutoff = now - DEDUP_WINDOW_MS
            seenIDs.entries.removeAll { it.value < cutoff }
        }
        return false
    }
}

File: ios/FestivalConnection/Mesh/Shared/PacketProcessor.swift
import Foundation
import CryptoKit

// MARK: - Packet Processor
// Central router that receives raw bytes from all transports (BLE, Multipeer, Nostr),
// deduplicates by message ID, decrements TTL and relays, and routes to handlers.

final class PacketProcessor: ObservableObject {
    static let shared = PacketProcessor()

    // Message deduplication cache
    private var seenIDs: [String: Date] = [:]
    private let dedupWindow: TimeInterval = 300
    private let maxCacheSize = 1000

    // Callbacks for routing by message type
    var onAnnounce: ((String, String, Date) -> Void)?         // peerKeyHex, nickname, timestamp
    var onMessage: ((String, String, String, Date) -> Void)?  // senderKeyHex, recipientKeyHex?, content, timestamp
    var onLeave: ((String) -> Void)?                          // peerKeyHex
    var onPaymentRequest: ((String, UInt64, String, String) -> Void)?  // senderHex, amount, invoice, description
    var onPaymentNotification: ((Data, UInt64, UInt8) -> Void)?        // hash, amount, direction

    private init() {}

    // MARK: - Process Incoming Data

    func receive(data: Data, fromTransport: TransportType) {
        let msgID = computeMessageID(data)
        guard !isDuplicate(msgID) else { return }

        guard let packet = CrowdSyncBinaryProtocol.decode(data) else { return }
        routePacket(packet)

        // Relay if TTL > 1
        if packet.ttl > 1 {
            let relayed = CrowdSyncPacket(
                type: packet.type, senderID: packet.senderID,
                recipientID: packet.recipientID, timestamp: packet.timestamp,
                payload: packet.payload, signature: packet.signature,
                ttl: packet.ttl - 1, version: packet.version
            )
            relay(relayed, excludingTransport: fromTransport)
        }
    }

    // MARK: - Route by Type

    private func routePacket(_ packet: CrowdSyncPacket) {
        let senderHex = packet.senderID.map { String(format: "%02x", $0) }.joined()
        let timestamp = Date(timeIntervalSince1970: Double(packet.timestamp) / 1000.0)

        guard let type = MessageType(rawValue: packet.type) else { return }

        switch type {
        case .announce:
            let nickname = String(data: packet.payload, encoding: .utf8) ?? "Peer \(senderHex.prefix(4).uppercased())"
            onAnnounce?(senderHex, nickname, timestamp)

        case .message:
            let content = String(data: packet.payload, encoding: .utf8) ?? ""
            let recipientHex = packet.recipientID.map { $0.map { String(format: "%02x", $0) }.joined() } ?? ""
            onMessage?(senderHex, recipientHex, content, timestamp)

        case .leave:
            onLeave?(senderHex)

        case .paymentRequest:
            if let decoded = PaymentPacketSerializer.decodePaymentRequest(packet.payload) {
                onPaymentRequest?(senderHex, decoded.amountSat, decoded.invoice, decoded.description)
            }

        case .paymentNotification:
            if let decoded = PaymentPacketSerializer.decodePaymentNotification(packet.payload) {
                onPaymentNotification?(decoded.paymentHash, decoded.amountSat, decoded.direction)
            }

        case .noiseHandshake, .noiseEncrypted, .fragment, .fileTransfer, .requestSync:
            break // Handled by specialized subsystems
        }
    }

    // MARK: - Relay to Other Transports

    private func relay(_ packet: CrowdSyncPacket, excludingTransport: TransportType) {
        guard let data = CrowdSyncBinaryProtocol.encode(packet) else { return }
        if excludingTransport != .ble {
            BLEService.shared.broadcast(data)
        }
        if excludingTransport != .multipeer {
            MultipeerTransportManager.shared.send(data)
        }
    }

    // MARK: - Send Outgoing

    func sendMessage(content: String, senderID: Data, recipientID: Data? = nil) {
        let payload = content.data(using: .utf8) ?? Data()
        let packet = CrowdSyncPacket(
            type: MessageType.message.rawValue,
            senderID: senderID,
            recipientID: recipientID,
            payload: payload,
            ttl: 7
        )
        guard let data = CrowdSyncBinaryProtocol.encode(packet) else { return }
        BLEService.shared.broadcast(data)
        MultipeerTransportManager.shared.send(data)
    }

    func sendPaymentRequest(invoice: String, amountSat: UInt64, description: String, senderID: Data, recipientID: Data) {
        let payload = PaymentPacketSerializer.encodePaymentRequest(invoice: invoice, amountSat: amountSat, description: description)
        let packet = CrowdSyncPacket(
            type: MessageType.paymentRequest.rawValue,
            senderID: senderID,
            recipientID: recipientID,
            payload: payload,
            ttl: 7
        )
        guard let data = CrowdSyncBinaryProtocol.encode(packet) else { return }
        BLEService.shared.broadcast(data)
        MultipeerTransportManager.shared.send(data)
    }

    // MARK: - Deduplication

    private func computeMessageID(_ data: Data) -> String {
        SHA256.hash(data: data).prefix(16).map { String(format: "%02x", $0) }.joined()
    }

    private func isDuplicate(_ id: String) -> Bool {
        if let seen = seenIDs[id], Date().timeIntervalSince(seen) < dedupWindow {
            return true
        }
        seenIDs[id] = Date()
        if seenIDs.count > maxCacheSize {
            let cutoff = Date().addingTimeInterval(-dedupWindow)
            seenIDs = seenIDs.filter { $0.value > cutoff }
        }
        return false
    }
}

// MARK: - Transport Type

enum TransportType {
    case ble
    case multipeer
    case nostr
}
